### DISCUSSION TITLE
Better snapshot formating

### DISCUSSION THREAD

**Discussion by Akiyamka** [avajs/ava#2903] [Ideas]
Current snapshot md file contain json that look like plain text.
It can be much more readable with ` ```js ` [around snap content](https://github.com/avajs/ava/blob/77623a5f32c139fa4ef7d316a4bbf7e172cc5acb/lib/snapshot-manager.js#L111).

| Current        | Can be           | 
| ------------- |:-------------:| 
| ![image](https://user-images.githubusercontent.com/6587585/143918574-3c5855d4-cde6-47f1-a6ed-146246429934.png)  | ![image](https://user-images.githubusercontent.com/6587585/143917843-0fd09bf7-9ed5-4e04-ab00-174fdbb6b2a3.png) |

**Comment by novemberborn** [comment]
The output isn't strictly JavaScript though, for example this is how a buffer is snapshotted:

    Buffer @Uint8Array [
      68656c6c 6f
    ]

**Reply by novemberborn** [reply]
Which happens to look OK enough in GitHub, but I'm not sure what other syntax bits the syntax highlighter may get weird about:


```js
    Buffer @Uint8Array [
      68656c6c 6f
    ]
```

**Reply by RandomByte** [reply]
Maybe the syntax (e.g. `"js"`) could be provided in the `t.snapshot()` call?

This would make snapshots (especially of JS) a lot more readable. We have a project where we generate JavaScript and syntax errors can happen. With some basic syntax highlighting in markdown you can spot this a lot easier.

**Reply by novemberborn** [reply]
@RandomByte are you snapshotting strings? They'll still be printed slightly differently in the Markdown.

The only way I see this working is a "raw" mode where the Markdown contains the string as-is, and then it would make sense to also provide a syntax tag.

**Reply by RandomByte** [reply]
@novemberborn Yes exactly. Strings of JavaScript code.

I guess you mean the backticks enclosing the string? They would indeed cancel out any JavaScript syntax highlighting, since the string simply becomes... a string in JavaScript.

**Currently I get something like this:**

````md
> Snapshot 1

    `console.log("Hello World");
    `
````

**My Expectation after somehow providing `js` syntax:**

````md
> Snapshot 1

    ```js
    console.log("Hello World");

    ```
````

Could this "raw" mode automatically engage if the developer provides a syntax tag?

**Reply by novemberborn** [reply]
Maybe an overload in `SnapshotAssertion`: https://github.com/avajs/ava/blob/a6f42ea472cfbe7047a0fb89ab0cd728b4e04ae1/types/assertions.d.cts#L320-L332)

```ts
snapshot(expected: string, options?: { formatAsCodeBlock: true | string }, message?: string)
```

This options object should only be accepted for string values, and must only contain `formatAsCodeBlock`. `string` values would be the syntax, and `true` would be the same as an empty string (but less weird).

When formatting, backticks would need to be escaped properly.

**Reply by RandomByte** [reply]
Nice proposal. Looks good to me!

**Reply by novemberborn** [reply]
@RandomByte would you want to open an issue for this? And, maybe, have a go at implementing it?

**Reply by novemberborn** [reply]
@RandomByte see https://github.com/avajs/ava/pull/3449.
